File: Binary Tree Preorder Traversal.java
// leetcode 144 problem
// 求二叉树的前序遍历序列

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */

// Recursive solution
class Solution {
    public List<Integer> preorderTraversal(TreeNode root) {
		List<Integer> pre = new LinkedList<Integer>();
		preHelper(root, pre);
		return pre;
	}
    
	public void preHelper(TreeNode root, List<Integer> pre) {
		if (root == null) return;
		pre.add(root.val);
		preHelper(root.left, pre);
		preHelper(root.right, pre);
	}
}

// iterative solution
public List<Integer> preorderTraversal(TreeNode root) {
    List<Integer> result = new ArrayList<>();
    Deque<TreeNode> stack = new ArrayDeque<>();
    TreeNode p = root;
    while(!stack.isEmpty() || p != null) {
        if(p != null) {
            stack.push(p);
            result.add(p.val);  // Add before going to children
            p = p.left;
        } else {
            TreeNode node = stack.pop();
            p = node.right;   
        }
    }
    return result;
}

File: Binary Tree Preorder Traversal.cpp
// Leetcode 144 problem
// 二叉树前序遍历（迭代解法）

class Solution {
public:
    vector<int> preorderTraversal(TreeNode *root) {
       vector<int> res;
       if (!root)
           return res;
       stack<TreeNode *> st;
       st.push(root);
       while (st.size()) {
           TreeNode *temp = st.top();
           st.pop();
           res.push_back(temp->val);
           if (temp->right)
               st.push(temp->right);
           if (temp->left)
               st.push(temp->left);
       }
       return res;
   }
};